File: swift/leet_code/988.swift
import Foundation

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */

struct Solution {
    /**
     * This function returns the smallest string that starts from a leaf of a binary tree and ends at its root.
     * The binary tree is mapped to a string by adding the ASCII value of 'a' to the node value and converting it to a character.
     * The function is optimized for speed.
     *
     * @param root: The root of the binary tree.
     * @param suffix: The suffix string to append to the current node's character. Default is an empty string.
     * @return: The smallest string that starts from a leaf and ends at the root.
     */
    @_optimize(speed)
    func smallestFromLeaf(_ root: TreeNode?, _ suffix: String = "") -> String {
        // If the root is nil, return an empty string
        guard let root else { return "" }

        // Convert the node value to a character and append the suffix
        let current = String(UnicodeScalar(UInt8(97 + root.val))) + suffix

        // If the node is a leaf, return the current string
        if root.left == nil, root.right == nil {
            return current
        }

        // Recursively call the function for the left and right children
        let left = root.left.map { smallestFromLeaf($0, current) }
        let right = root.right.map { smallestFromLeaf($0, current) }

        // Return the smallest string between the left and right children
        // If one of them is nil, return the other
        if let left, let right {
            return left < right ? left : right
        } else if left == nil {
            return right ?? ""
        } else {
            return left ?? ""
        }
    }
}

File: go/leet_code/988.go
// Package leet_code provides solutions for LeetCode problems.
package leet_code

// Importing the fmt package for string formatting
import "fmt"

// TreeNode represents a node in a binary tree.
// Each node contains an integer value, a pointer to the left child node, and a pointer to the right child node.
type TreeNode struct {
	Val   int       // The value of the node.
	Left  *TreeNode // Pointer to the left child node.
	Right *TreeNode // Pointer to the right child node.
}

// Global variable to store the result string
var result string

// travel is a helper function that traverses the binary tree in a pre-order manner (root, left, right).
// It constructs strings from the node values and keeps track of the smallest string seen so far.
//
// Parameters:
//  - node: The current node being visited.
//  - s: The string constructed so far.
//
// Returns:
//  - None. The result is stored in the global variable 'result'.
//
// Error Handling:
//  - If the node is nil, the function returns without doing anything.
//
// Example Usage:
//  root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
//  s := ""
//  travel(root, s) // The result is stored in the global variable 'result'.
func travel(node *TreeNode, s string) {
	if node == nil {
		return
	}

	// Prepend the current node's value to the string
	s = fmt.Sprintf("%c", node.Val+97) + s

	// If the node is a leaf node, update the result
	if node.Left == nil && node.Right == nil {
		if result == "" {
			result = s
		} else {
			result = min(result, s)
		}
		return
	}

	// Recursively call the function for the left and right child nodes
	if node.Left != nil {
		travel(node.Left, s)
	}
	if node.Right != nil {
		travel(node.Right, s)
	}
}

// smallestFromLeaf finds the smallest string that starts from a leaf of a binary tree and ends at the root.
// The string is constructed by concatenating the characters corresponding to the node values in the path from the leaf to the root.
// The characters are determined by adding 97 to the node value to get the corresponding lowercase ASCII character.
//
// Parameters:
//  - root: The root node of the binary tree.
//
// Returns:
//  - The smallest string that starts from a leaf and ends at the root.
//
// Error Handling:
//  - If the root is nil, the function returns an empty string.
//
// Example Usage:
//  root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
//  smallestString := smallestFromLeaf(root) // smallestString now contains the smallest string from leaf to root.
func smallestFromLeaf(root *TreeNode) string {
	// Reset the result
	result = ""

	// Call the helper function to traverse the tree and find the smallest string
	travel(root, "")

	// Return the result
	return result
}
